artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireConnection.java

```diff
@@ -698,6 +700,21 @@ public void fail(ActiveMQException me, String message) {
       shutdown(true);
    }
 
+   private void returnReferencesFromAnyPendingLocalTransactions() {
+      for (Transaction tx : txMap.values()) {
+         final AMQSession amqSession = (AMQSession) tx.getProtocolData();
+         if (amqSession != null) {
+            amqSession.getCoreSession().resetTX(tx);
+            try {
+               commandProcessorInstance.returnReferences(tx, amqSession);
```

Comment:
this whole code here would be simply:

```java
   private void returnReferencesFromAnyPendingLocalTransactions() {
      for (Transaction tx : txMap.values()) {
         try {
            tx.rollback();
         } catch (Exception e) {
            logger.warn(e.getMessage());
         }
      }
   }
```

I don't know the reason for for returnReferences command? I think it's because open wire can redeliver messages locally without a roundtrip to the server, hence you need to make a distinction.

but on this case, a simple rollback on the transaction would suffice.

Reply:
nice, that is simpler, I found rollbackIfPossible that will ignore pending xa. It fixes the issue. thanks. Will push an update with the change in a bit.